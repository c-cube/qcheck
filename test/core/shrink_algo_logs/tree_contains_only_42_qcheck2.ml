open QCheck2

module IntTree = struct
  type tree = Leaf of int | Node of tree * tree

  let leaf x = Leaf x
  let node x y = Node (x,y)

  let rec depth = function
    | Leaf _ -> 1
    | Node (x, y) -> 1 + max (depth x) (depth y)

  let rec print_tree = function
    | Leaf x -> Printf.sprintf "Leaf %d" x
    | Node (x, y) -> Printf.sprintf "Node (%s, %s)" (print_tree x) (print_tree y)

  let gen_tree = Gen.(sized @@ fix
                        (fun self n -> match n with
                           | 0 -> map leaf nat
                           | n ->
                             frequency
                               [1, map leaf nat;
                                2, map2 node (self (n/2)) (self (n/2))]
                        ))

  let rec contains_only_n tree n = match tree with
    | Leaf n' -> n = n'
    | Node (x, y) -> contains_only_n x n && contains_only_n y n
end

let tree_contains_only_42 =
  Test.make ~name:"tree contains only 42"
    IntTree.gen_tree
    (Log.shrinks IntTree.print_tree @@ fun tree -> IntTree.contains_only_n tree 42)

let rand = Random.State.make [|1234|];;
try Test.check_exn ~rand tree_contains_only_42 with Test.Test_fail _ -> ()
