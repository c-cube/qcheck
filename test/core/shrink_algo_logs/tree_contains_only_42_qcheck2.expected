fails Node (Leaf 174, Leaf 31)
fails Leaf 9
fails Leaf 0
